service/raft/raft_group0.cc

```diff
@@ -230,135 +236,228 @@ future<> raft_group0::abort() {
+            // FIXME if we crash now or after getting added to the config but before storing group 0 ID,
+            // we'll end with a bootstrapped server that possibly added some entries, but we won't remember that we have such a server
+            // after we restart. Then we'll call `persistence.bootstrap` again after restart which will overwrite our snapshot, leading to
+            // possibly incorrect state. One way of handling this may be changing `persistence.bootstrap` so it checks if any persistent
+            // state is present, and if it is, do nothing.
         }
-        if (server->get_configuration().contains(my_addr.id)) {
-            // True if we started a new group or completed a configuration change
-            // initiated earlier.
-            rslog.trace("server {} already in group as {}", my_addr.id,
-                    server->get_configuration().can_vote(my_addr.id)? "voter" : "non-voter");
+
+        assert(server);
+        if (server->get_configuration().contains(my_id)) {
+            // True if we started a new group or completed a configuration change initiated earlier.
+            rslog.info("server {} already in group 0 (id {}) as {}", group0_id, my_id,
+                    server->get_configuration().can_vote(my_id)? "voter" : "non-voter");
             break;
         }
-        std::vector<raft::server_address> add_set;
-        add_set.push_back(my_addr);
-        auto pause_shutdown = _shutdown_gate.hold();
+
+        // XXX is 1 second timeout enough in a multi-dc cluster?
         auto timeout = db::timeout_clock::now() + std::chrono::milliseconds{1000};
         netw::msg_addr peer(raft_addr_to_inet_addr(g0_info.addr));
         try {
-            co_await ser::group0_rpc_verbs::send_group0_modify_config(&_ms, peer, timeout, group0_id, add_set, {});
+            // TODO abort?
+            co_await ser::group0_rpc_verbs::send_group0_modify_config(&_ms, peer, timeout, group0_id, {raft::server_address{
+                    .id = my_id,
+                    .can_vote = as_voter,
+                    .info = my_info,
+                }}, {});
             break;
+        } catch (abort_requested_exception&) {
+            throw;
         } catch (std::runtime_error& e) {
             // Retry
             rslog.error("failed to modify config at peer {}: {}", g0_info.addr.id, e);
         }
+
         // Try again after a pause
         co_await seastar::sleep_abortable(std::chrono::milliseconds{1000}, _abort_source);
     }
     co_await db::system_keyspace::set_raft_group0_id(group0_id.id);
     // Allow peer_exchange() RPC to access group 0 only after group0_id is persisted.
 
     _group0 = group0_id;
-    rslog.info("{} joined group 0 with id {}", my_addr.id, group0_id);
+    rslog.info("{} joined group 0 with id {}", my_id, group0_id);
 }
 
-future<> raft_group0::become_voter() {
+future<> raft_group0::bootstrap(const std::unordered_set<gms::inet_address>& initial_contact_nodes) {
```

Comment:
I don't like this rename. First of all, I like names to be easy to grep. This name isn't. Second, this isn't just bootstrap.

This is a catch all for boostrapping a new node, boostrapping with raft off, restarting or upgrading. The old name was join_group0(), because its post-condition was that the cluster is part of group0. Later we added the disable option and it slowly started to turn into a mess.

Maybe this method should be inlined into the storage service, since it implements the general booting logic. 

I think important context is missing. We never say here that we do not allow adding nodes to an existing cluster, so when we boot a new node we always enable SUPPORTS_RAFT, and then the persistent features machinery does the rest for us - won't let the node proceed with booting if other nodes in the cluster don't advertise the feature.

Understanding this is key to understanding how the rest of the method works.

Reply:
> I don't like this rename. First of all, I like names to be easy to grep. This name isn't. Second, this isn't just bootstrap.

Maybe. I would like a better name too. Maybe something like `setup_group0` - easy to grep (although `group0.*bootstrap` *is* easy to grep).

Describing everything what this function does cannot be done with a short name; similarly `storage_service::join_cluster` really does a lot of things, because joining a cluster involves a lot of things. My thinking was that bootstrapping group 0 involves a lot of things.

I think I need to write a comment above `bootstrap` declaration that summarizes what it does.

> Maybe this method should be inlined into the storage service, since it implements the general booting logic.

I disagree. It implements the booting logic *specific to Raft and group 0*: it joins group 0 or prepares for upgrading the cluster to use to group 0. That's why I think the code should be inside a group 0 specific module.

I think the consensus among maintainers is that we should reduce `join_cluster` to a sequence of very high-level steps that call to different modules which need to do something during cluster joining: synchronize features, update gossip states, establish cdc generation, setup group 0...

For me the picture is as follows: we have a bunch of modules (features, gossip, cdc, group 0) which all have something to say during the joining cluster procedure. There is an acyclic graph of dependencies between the steps of those modules: e.g. cdc assumes that features are checked when it's invoked. The point of `storage_service::join_cluster` is is to fix a certain topological order on this graph of dependencies and call each modules' steps in this order.

Reply:
> I think important context is missing. We never say here that we do not allow adding nodes to an existing cluster, so when we boot a new node we always enable SUPPORTS_RAFT, and then the persistent features machinery does the rest for us - won't let the node proceed with booting if other nodes in the cluster don't advertise the feature.

I'm not sure what you're driving at here.
Are you talking about the `is_enabled` check?
The check does not check SUPPORTS_RAFT feature, it checks the **local experimental feature**. If the local experimental feature is off, bootstrapping group0 is a no-op.